hibernate-core/src/main/java/org/hibernate/event/internal/DefaultRefreshEventListener.java

```diff
@@ -65,21 +65,42 @@ public void onRefresh(RefreshEvent event, RefreshContext refreshedAlready) {
 		final Object object = event.getObject();
 		if ( persistenceContext.reassociateIfUninitializedProxy( object ) ) {
 			final boolean isTransient = isTransient( event, source, object );
+			// If refreshAlready is not empty then the refresh is the result of a
+			// cascade refresh and the refresh of the parent will take care of initializing the lazy
+			// entity and setting the correct lock on it, this is needed only when the refresh is called directly on a lazy entity
+			if ( refreshedAlready.isEmpty() ) {
```

Comment:
```suggestion
			if ( refreshedAlready.add( entity ) ) {
```

Reply:
we want to call it only when is not a call from a refresh cascade operation, not sure if there is a better way than checking  `refreshedAlready` is empty,

Reply:
I'm a bit confused about the comment. This method is also called through cascading `org.hibernate.engine.spi.CascadingActions.REFRESH#cascade`, but the `refreshedAlready` set might be non-empty.

Reply:
Ohh, I see now. Ok, forget my comment.